modules/tensor_mechanics/src/materials/CalculateStrainLagrangianKernel.C

```diff
@@ -0,0 +1,248 @@
+#include "CalculateStrainLagrangianKernel.h"
+
+registerMooseObject("TensorMechanicsApp", CalculateStrainLagrangianKernel);
+
+InputParameters
+CalculateStrainLagrangianKernel::validParams()
+{
+  InputParameters params = Material::validParams();
+
+  params.addRequiredCoupledVar("displacements", "Displacement variables");
+  params.addParam<bool>(
+      "kernel_large_kinematics", false, "Use large displacement kinematics in the kernel.");
+  params.addParam<bool>("material_large_kinematics",
+                        false,
+                        "If true the material returns the Cauchy stress"
+                        " and the derivative with respect to the spatial "
+                        "velocity gradient directly.");
+  params.addParam<std::vector<MaterialPropertyName>>("eigenstrain_names",
+                                                     "List of eigenstrains to account for.");
+  params.addCoupledVar("macro_gradient", "Optional scalar field with the macro gradient");
+
+  // We rely on this *not* having use_displaced mesh on
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  return params;
+}
+
+CalculateStrainLagrangianKernel::CalculateStrainLagrangianKernel(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Material>(parameters),
+    _ndisp(coupledComponents("displacements")),
+    _disp(3),
+    _grad_disp(3),
+    _ld_kernel(getParam<bool>("kernel_large_kinematics")),
+    _ld_material(getParam<bool>("material_large_kinematics")),
+    _eigenstrain_names(getParam<std::vector<MaterialPropertyName>>("eigenstrain_names")),
+    _eigenstrains(_eigenstrain_names.size()),
+    _eigenstrains_old(_eigenstrain_names.size()),
+    _total_strain(declareProperty<RankTwoTensor>("total_strain")),
+    _total_strain_old(getMaterialPropertyOld<RankTwoTensor>("total_strain")),
+    _mechanical_strain(declareProperty<RankTwoTensor>("mechanical_strain")),
+    _mechanical_strain_old(getMaterialPropertyOld<RankTwoTensor>("mechanical_strain")),
+    _strain_increment(declareProperty<RankTwoTensor>("strain_increment")),
+    _rotation_increment(declareProperty<RankTwoTensor>("rotation_increment")),
+    _def_grad(declareProperty<RankTwoTensor>("deformation_gradient")),
+    _def_grad_old(getMaterialPropertyOld<RankTwoTensor>("deformation_gradient")),
+    _df(declareProperty<RankTwoTensor>("inc_def_grad")),
```

Comment:
If I understand it correctly, this shouldn't be called the inc_def_grad, it is actually the inverse increment.